File: Algorithms/253_Meeting_Rooms_II/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

class Solution {

    public int minMeetingRooms(int[][] intervals) {
        if (intervals == null || intervals.length == 0) {
            return 0;
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        for (int[] interval : intervals) {
            pq.offer(interval);
        }

        List<List<int[]>> reserved = new ArrayList<>() {{
            add(new ArrayList<>() {{
                add(pq.poll());
            }});
        }};

        loop: while (!pq.isEmpty()) {
            reserved.sort(Comparator.comparing(a -> a.get(a.size() - 1)[1]));
            int[] itv = pq.poll();
            for (List<int[]> itvs : reserved) {
                if (itvs.get(itvs.size() - 1)[1] <= itv[0]) {
                    itvs.add(itv);
                    continue loop;
                }
            }
            reserved.add(new ArrayList<>() {{
                add(itv);
            }});
        }

        return reserved.size();
    }
}

File: Algorithms/253_Meeting_Rooms_II/solution.go
package main

import "sort"

// 현재 우측점을 기준으로 최대 겹치는 부분을 탐색
func minMeetingRooms(intervals [][]int) int {
    starts, ends := make([]int, 0, len(intervals)), make([]int, 0, len(intervals))
    for _, itv := range intervals {
        starts, ends = append(starts, itv[0]), append(ends, itv[1])
    }
    sort.Ints(starts)
    sort.Ints(ends)

    overlap, endPos := 0, 0
    for _, start := range starts {
        if start < ends[endPos] {
            overlap++
        } else {
            endPos++
        }
    }

    return overlap
}
